solidity/contracts/TokenStaking.sol

```diff
@@ -619,13 +671,27 @@ contract TokenStaking is StakeDelegatable {
      * @return True if has enough active stake to participate in the network,
      * false otherwise.
      */
-    function hasMinimumStake(
-        address staker,
-        address operatorContract
-    ) public view returns(bool) {
+    function hasMinimumStake(address staker, address operatorContract)
+        public
+        view
+        returns (bool)
+    {
         return activeStake(staker, operatorContract) >= minimumStake();
     }
 
+    /// @notice Grants the contract the same authority as `msg.sender` has.
+    /// @dev `msg.sender` has to be an approved operator contract.
+    /// Later, the claimant can slash, seize, place locks etc.
+    /// on operators that have authorized the `msg.sender`.
+    /// If the `msg.sender` is disabled with the panic button,
+    /// any recipients of delegated authority from it will also be disabled.
+    function delegateAuthority(address delegateContract)
+        public
+        onlyApprovedOperatorContract(msg.sender)
+    {
+        delegatedAuthority[delegateContract] = msg.sender;
```

Comment:
I considered an one-step process when making `claimDelegatedAuthority` but it violates some security principles. There is no check that the target of `delegateAuthority` is actually trying to get delegated authority from the caller, meaning that any approved contract A can delegate its authority to any other contract B, preventing B from functioning unless operators authorize A. This is particularly troublesome if a permanently approved contract turns out to have a bug which lets it delegate authority to arbitrary addresses, as it can shut down the entire network.

I settled on `claimDelegatedAuthority` because keep factories call `initialize` after creating the proxied keep, thus giving them time to record the keep's address before it tries to claim delegated authority. An alternative method would be to have the recipient of delegated authority define which contract it wishes to get delegated authority from, and check this in the call to `delegateAuthority`.

Reply:
Good point. I went through the code once again, made some additional tests and I think we are good with what we have now. Even if the interaction here is probably more intuitive for the factory, the *claim* version is more robust. I'll close this one without merging.